File: sensorhub-backend/src/main/java/com/sensorhub/platform/controller/SensorWorkspaceController.java
package com.sensorhub.platform.controller;

import com.sensorhub.platform.annotation.AuthCheck;
import com.sensorhub.platform.common.BaseResponse;
import com.sensorhub.platform.common.ResultUtils;
import com.sensorhub.platform.constant.UserConstant;
import com.sensorhub.platform.model.vo.SensorWorkspaceCatalogSyncResultVO;
import com.sensorhub.platform.model.vo.SensorWorkspaceInterfaceVO;
import com.sensorhub.platform.service.SensorWorkspaceService;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.annotation.Resource;
import java.util.List;
import java.util.Map;

/**
 * Local sensor workspace controller.
 */
@RestController
@RequestMapping("/sensorWorkspace")
public class SensorWorkspaceController {

    @Resource
    private SensorWorkspaceService sensorWorkspaceService;

    @GetMapping("/catalog")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<List<SensorWorkspaceInterfaceVO>> getCatalog() {
        return ResultUtils.success(sensorWorkspaceService.listNorthboundInterfaces());
    }

    @GetMapping("/overview")
    public BaseResponse<Map<String, Object>> getOverview() {
        return ResultUtils.success(sensorWorkspaceService.getWorkspaceOverview());
    }

    @GetMapping("/devices")
    public BaseResponse<List<Map<String, Object>>> getDevices() {
        return ResultUtils.success(sensorWorkspaceService.listSensorDevices());
    }

    @GetMapping("/realtimeWorkspace")
    public BaseResponse<Map<String, Object>> getRealtimeWorkspace() {
        return ResultUtils.success(sensorWorkspaceService.getRealtimeWorkspace());
    }

    @PostMapping("/devices")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Map<String, Object>> createDevice(@RequestBody Map<String, Object> devicePayload) {
        return ResultUtils.success(sensorWorkspaceService.createSensorDevice(devicePayload));
    }

    @PutMapping("/devices/{deviceId}")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Map<String, Object>> updateDevice(
            @PathVariable("deviceId") String deviceId,
            @RequestBody Map<String, Object> devicePayload
    ) {
        return ResultUtils.success(sensorWorkspaceService.updateSensorDevice(deviceId, devicePayload));
    }

    @DeleteMapping("/devices/{deviceId}")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Map<String, Object>> deleteDevice(@PathVariable("deviceId") String deviceId) {
        return ResultUtils.success(sensorWorkspaceService.deleteSensorDevice(deviceId));
    }

    @PostMapping("/devices/test")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Map<String, Object>> testDevice(@RequestBody Map<String, Object> devicePayload) {
        return ResultUtils.success(sensorWorkspaceService.testSensorDevice(devicePayload));
    }

    @GetMapping("/realtime")
    public BaseResponse<Map<String, Object>> getRealtimeSnapshot(
            @RequestParam(value = "limit", required = false, defaultValue = "5") Integer limit
    ) {
        return ResultUtils.success(sensorWorkspaceService.getRealtimeSnapshot(limit));
    }

    @GetMapping("/search")
    public BaseResponse<Map<String, Object>> searchDataset(
            @RequestParam("dataset") String dataset,
            @RequestParam(value = "deviceToken", required = false) String deviceToken,
            @RequestParam(value = "startDate", required = false) String startDate,
            @RequestParam(value = "endDate", required = false) String endDate,
            @RequestParam(value = "limit", required = false, defaultValue = "20") Integer limit
    ) {
        return ResultUtils.success(
                sensorWorkspaceService.searchDataset(dataset, deviceToken, startDate, endDate, limit)
        );
    }

    @PostMapping("/sync")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<SensorWorkspaceCatalogSyncResultVO> syncCatalog() {
        return ResultUtils.success(sensorWorkspaceService.syncNorthboundInterfaces());
    }
}

File: sensorhub-frontend/src/services/sensorhub-backend/sensorWorkspaceController.ts
// @ts-ignore
/* eslint-disable */
import { request } from '@umijs/max';

/** getCatalog GET /api/sensorWorkspace/catalog */
export async function getSensorWorkspaceCatalogUsingGET(options?: { [key: string]: any }) {
  return request<any>('/api/sensorWorkspace/catalog', {
    method: 'GET',
    ...(options || {}),
  });
}

/** syncCatalog POST /api/sensorWorkspace/sync */
export async function syncSensorWorkspaceCatalogUsingPOST(options?: { [key: string]: any }) {
  return request<any>('/api/sensorWorkspace/sync', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    ...(options || {}),
  });
}

/** getOverview GET /api/sensorWorkspace/overview */
export async function getSensorWorkspaceOverviewUsingGET(options?: { [key: string]: any }) {
  return request<any>('/api/sensorWorkspace/overview', {
    method: 'GET',
    ...(options || {}),
  });
}

/** getDevices GET /api/sensorWorkspace/devices */
export async function listSensorWorkspaceDevicesUsingGET(options?: { [key: string]: any }) {
  return request<any>('/api/sensorWorkspace/devices', {
    method: 'GET',
    ...(options || {}),
  });
}

/** getRealtimeWorkspace GET /api/sensorWorkspace/realtimeWorkspace */
export async function getSensorWorkspaceRealtimeWorkspaceUsingGET(options?: {
  [key: string]: any;
}) {
  return request<any>('/api/sensorWorkspace/realtimeWorkspace', {
    method: 'GET',
    ...(options || {}),
  });
}

/** createDevice POST /api/sensorWorkspace/devices */
export async function createSensorWorkspaceDeviceUsingPOST(
  body: Record<string, any>,
  options?: { [key: string]: any },
) {
  return request<any>('/api/sensorWorkspace/devices', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** updateDevice PUT /api/sensorWorkspace/devices/${param0} */
export async function updateSensorWorkspaceDeviceUsingPUT(
  params: {
    deviceId: string;
  },
  body: Record<string, any>,
  options?: { [key: string]: any },
) {
  const { deviceId: param0, ...queryParams } = params;
  return request<any>(`/api/sensorWorkspace/devices/${param0}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    params: { ...queryParams },
    data: body,
    ...(options || {}),
  });
}

/** deleteDevice DELETE /api/sensorWorkspace/devices/${param0} */
export async function deleteSensorWorkspaceDeviceUsingDELETE(
  params: {
    deviceId: string;
  },
  options?: { [key: string]: any },
) {
  const { deviceId: param0, ...queryParams } = params;
  return request<any>(`/api/sensorWorkspace/devices/${param0}`, {
    method: 'DELETE',
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** testDevice POST /api/sensorWorkspace/devices/test */
export async function testSensorWorkspaceDeviceUsingPOST(
  body: Record<string, any>,
  options?: { [key: string]: any },
) {
  return request<any>('/api/sensorWorkspace/devices/test', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** getRealtimeSnapshot GET /api/sensorWorkspace/realtime */
export async function getSensorWorkspaceRealtimeUsingGET(
  params: {
    limit?: number;
  },
  options?: { [key: string]: any },
) {
  return request<any>('/api/sensorWorkspace/realtime', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** searchDataset GET /api/sensorWorkspace/search */
export async function searchSensorWorkspaceDatasetUsingGET(
  params: {
    dataset: string;
    deviceToken?: string;
    startDate?: string;
    endDate?: string;
    limit?: number;
  },
  options?: { [key: string]: any },
) {
  return request<any>('/api/sensorWorkspace/search', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}
